timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/watch/LockWatchingServiceImpl.java

```diff
@@ -147,16 +156,42 @@ private void logLockWatchEvent(LockWatches newWatches) {
 
     private void runIfDescriptorsMatchLockWatches(
             Set<LockDescriptor> unfiltered, Consumer<Set<LockDescriptor>> consumer) {
+        runIfDescriptorsMatchLockWatchesWithMetadata(
+                unfiltered, Optional.empty(), (lockDescriptors, unused) -> consumer.accept(lockDescriptors));
+    }
+
+    private void runIfDescriptorsMatchLockWatchesWithMetadata(
+            Set<LockDescriptor> unfilteredLocks,
+            Optional<LockRequestMetadata> unfilteredMetadata,
+            BiConsumer<Set<LockDescriptor>, Optional<LockRequestMetadata>> biConsumer) {
         watchesLock.readLock().lock();
         try {
             RangeSet<LockDescriptor> ranges = watches.get().ranges();
-            Set<LockDescriptor> filtered =
-                    unfiltered.stream().filter(ranges::contains).collect(Collectors.toSet());
-            if (!filtered.isEmpty()) {
-                consumer.accept(filtered);
+            Set<LockDescriptor> filteredLocks =
+                    unfilteredLocks.stream().filter(ranges::contains).collect(Collectors.toSet());
+            if (filteredLocks.isEmpty()) {
+                return;
             }
+            // Filtering metadata after deciding if we should even proceed might save us some computation
+            biConsumer.accept(filteredLocks, filterMetadataBasedOnFilteredLocks(filteredLocks, unfilteredMetadata));
         } finally {
             watchesLock.readLock().unlock();
         }
     }
+
+    // For an efficient encoding, we enforce that metadata is never attached to a lock descriptor that is not contained
+    // in the original request, so filtering metadata based on the already filtered locks is sufficient.
+    // It is also cheaper than calling RangeSet::contains.
+    private static Optional<LockRequestMetadata> filterMetadataBasedOnFilteredLocks(
+            Set<LockDescriptor> filteredLocks, Optional<LockRequestMetadata> unfilteredMetadata) {
+        return unfilteredMetadata
```

Comment:
The other question I have is how can we defend against this? I.e. if this invariant breaks, how will we find out?

Reply:
Our wire format (index encoding) will enforce that this invariant is upheld.
However, if we want an explicit check to abstract from the wire format, we can do so when checking the maximum size of the metadata. As mentioned in the PR description, I wanted to do this in a separate PR.
Question is if we even care if the invariant might not hold when receiving metadata in TimeLock, since it will always hold after filtering. Maybe we can loosen the wording here from `enforce` to `expect`.

Reply:
Yeah more so of the point, is that, we have an invariant enforced _elsewhere_ in our code; if that breaks/changes, how would we find out about it here?

Reply:
As discussed, I have tackled this with an assert and a unit test that listens for that assert. If somebody violates the invariant and tries to pass metadata for an unknown lock through timelock, this assert will throw an error.